conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+                param_norm = p.grad.data.norm(norm_type)
+            total_norm += param_norm ** norm_type
+        total_norm = total_norm ** (1. / norm_type)
+    clip_coef = float(max_norm / (total_norm + 1e-6))
+    if clip_coef < 1:
+        for p in parameters:
+            p.grad.data.mul_(clip_coef)
+    return total_norm
+
+
+
+class SGD_Sparse(optim.SGD):
+    """
+    Just like optim.SGD except that this class handles non-zero 
+    weight decay and momentum even when some of the parameters 
+    being optimized have sparse gradient storage.
+    
+    NOTE:  Currently there seem to be gpu memory leaks when using 
+    Nesterov momentum; use that at your own risk.
+    """
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+                d_p_is_sparse = (d_p.layout == torch.sparse_coo)
+                if weight_decay != 0:
+                    if not d_p_is_sparse:
+                        d_p.add_(weight_decay, p.data)
+                    else:
+                        # Update the gradient via weight decay only at
+                        # locations where it is already non-zero; this
+                        # isn't quite the same as actual weight decay but
+                        # it avoids ballooning the memory consumption.
+                        d_p = d_p.coalesce()
+                        p.grad.data = d_p
+                        if len(d_p._indices().size()) < 2: # empty sparse tensor
+                            pass
+                        else:
+                            # Find the indices of the nonzero entries in the
+                            # gradient, assuming row-major storage order, by
+                            # finding the index of each nonzero subtensor's
+                            # first entry and adding offsets to the other
+                            # entries.
+                            assert p.data.size() == d_p.size()
+                            n_sparse_dims, n_values = d_p._indices().size()
+                            values_shape = tuple(d_p._values().size())
+                            assert len(values_shape) > 0
+                            assert n_values == values_shape[0]
+                            value_size = int(np.prod(np.array(
+                                values_shape[1:])))
+                            factors = np.array(d_p.size())
```

Comment:
Again, I tried to explain further in comments committed before removing the class.  Since the class is gone now, I won't explain further (unless you ask).